packages/worker/src/workflow/vm.ts

```diff
@@ -0,0 +1,166 @@
+   *
+   * Overridable for test purposes.
+   */
+  protected injectConsole(context: vm.Context, info: WorkflowInfo): void {
+    context.console = {
+      log: (...args: any[]) => {
+        // info.isReplaying is mutated by the Workflow class on activation
+        if (info.isReplaying) return;
+        console.log(`[${info.workflowType}(${info.workflowId})]`, ...args);
+      },
+    };
+  }
+
+  /**
+   * Create a new instance, isolates and pre-compiled scripts are generated here
+   */
+  public static async create<T extends typeof VMWorkflowCreator>(
+    this: T,
+    code: string,
+    isolateExecutionTimeoutMs: number
+  ): Promise<InstanceType<T>> {
+    const script = new vm.Script(code, { filename: 'workflow-isolate' });
+    return new this(script, isolateExecutionTimeoutMs) as InstanceType<T>;
+  }
+
+  public async destroy(): Promise<void> {
+    delete this.script;
+  }
+}
+
+type WorkflowModule = typeof internals;
+
+export class VMWorkflow implements Workflow {
+  constructor(
+    public readonly info: WorkflowInfo,
+    protected context: vm.Context | undefined,
+    readonly workflowModule: WorkflowModule,
+    public readonly isolateExecutionTimeoutMs: number
+  ) {}
+
+  async getAndResetSinkCalls(): Promise<SinkCall[]> {
+    return this.workflowModule.getAndResetSinkCalls();
+  }
+
+  /**
+   * Inject a function into the isolate context global scope
+   *
+   * @param path name of global variable to inject the function as (e.g. `console.log`)
+   * @param fn function to inject into the isolate
+   */
+  public async injectGlobal(key: string, val: unknown): Promise<void> {
+    if (this.context === undefined) {
+      throw new IllegalStateError('Workflow isolate context uninitialized');
+    }
+    this.context[key] = val;
+  }
+
+  public async activate(activation: coresdk.workflow_activation.IWFActivation): Promise<Uint8Array> {
+    if (this.context === undefined) {
+      throw new IllegalStateError('Workflow isolate context uninitialized');
+    }
+    this.info.isReplaying = activation.isReplaying ?? false;
+    if (!activation.jobs) {
+      throw new Error('Expected workflow activation jobs to be defined');
+    }
+
+    // Job processing order
+    // 1. patch notifications
+    // 2. signals
+    // 3. anything left except for queries
+    // 4. queries
+    const [patches, nonPatches] = partition(activation.jobs, ({ notifyHasPatch }) => notifyHasPatch != null);
+    const [signals, nonSignals] = partition(nonPatches, ({ signalWorkflow }) => signalWorkflow != null);
+    const [queries, rest] = partition(nonSignals, ({ queryWorkflow }) => queryWorkflow != null);
+    let batchIndex = 0;
+
+    // Loop and invoke each batch and wait for microtasks to complete.
+    // This is done outside of the isolate because when we used isolated-vm we couldn't wait for microtasks from inside the isolate, not relevant anymore.
+    for (const jobs of [patches, signals, rest, queries]) {
```

Comment:
Just thinking out loud here - what if you passed the whole job list in the message to activate rather than one at a time? Might make for some significant easy savings?